File: .github/workflows/sync.py
# /// script
# requires-python = "==3.10"
# dependencies = [
#   "requests",
# ]
# ///
# ruff: noqa: T201
import os
import subprocess
import sys
import time
import uuid

import requests


def main():
    GITHUB_TOKEN = os.environ.get("GITHUB_TOKEN")
    if not GITHUB_TOKEN:
        print("GITHUB_TOKEN not found in environment variables", file=sys.stderr)
        sys.exit(1)

    OWNER = "mlflow"
    REPO = "mlflow"
    MLFLOW_3_BRANCH_NAME = "mlflow-3"
    PR_TITLE = "Sync with master"
    PR_BRANCH_NAME = f"sync-{uuid.uuid4()}"

    # Exit if there is already a PR with the same title
    def iter_pull_requests():
        per_page = 100
        page = 1
        while True:
            resp = requests.get(
                f"https://api.github.com/repos/{OWNER}/{REPO}/pulls",
                headers={"Authorization": f"token {GITHUB_TOKEN}"},
                params={
                    "state": "open",
                    "base": MLFLOW_3_BRANCH_NAME,
                    "per_page": per_page,
                    "page": page,
                },
            )
            resp.raise_for_status()
            prs = resp.json()
            yield from prs
            if len(prs) < per_page:
                break
            page += 1

    # Fetch master and mlflow-3 branches
    subprocess.check_call(["git", "config", "user.name", "mlflow-automation"])
    subprocess.check_call(
        ["git", "config", "user.email", "mlflow-automation@users.noreply.github.com"]
    )
    subprocess.check_call(["git", "fetch", "origin", "master"])
    subprocess.check_call(["git", "fetch", "origin", MLFLOW_3_BRANCH_NAME])

    # Sync mlflow-3 branch with master
    subprocess.check_call(
        ["git", "checkout", "-b", PR_BRANCH_NAME, f"origin/{MLFLOW_3_BRANCH_NAME}"]
    )
    prc = subprocess.run(
        ["git", "merge", "origin/master"],
        text=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
    )
    print(prc.stdout)
    if prc.returncode != 0:
        print(
            (
                "Merge failed, possibly due to conflicts. "
                "Follow the instructions in `sync.md` to resolve the conflicts "
                "and create a PR manually."
            ),
            file=sys.stderr,
        )
        sys.exit(1)
    elif "Already up to date" in prc.stdout:
        print("Branch is already up to date with master, no changes to sync")
        sys.exit(0)

    if existing_pr := next((pr for pr in iter_pull_requests() if pr["title"] == PR_TITLE), None):
        head_ref = existing_pr["head"]["ref"]
        assert head_ref.startswith("sync-"), f"Unexpected head ref: {head_ref}"
        subprocess.check_call(
            [
                "git",
                "push",
                "-f",
                "origin",
                f"{PR_BRANCH_NAME}:{head_ref}",
            ]
        )
        print(f"Updated existing PR: {existing_pr['html_url']}")
        sys.exit(0)
    else:
        subprocess.check_call(
            [
                "git",
                "push",
                "origin",
                PR_BRANCH_NAME,
            ]
        )

    # PR creation right after the push sometimes fails with 422,
    # so retry up to 5 times with an exponential backoff
    for i in range(5):
        pr = requests.post(
            f"https://api.github.com/repos/{OWNER}/{REPO}/pulls",
            headers={"Authorization": f"token {GITHUB_TOKEN}"},
            json={
                "title": PR_TITLE,
                "head": PR_BRANCH_NAME,
                "base": MLFLOW_3_BRANCH_NAME,
                "body": "This PR was created automatically by the sync workflow.",
            },
        )
        try:
            pr.raise_for_status()
        except requests.HTTPError as e:
            if e.response.status_code == 422:
                print(
                    f"Failed to create PR (error: {e}), retrying in {2**i} seconds", file=sys.stderr
                )
                time.sleep(2**i)
                continue
            raise
        else:
            pr_data = pr.json()
            print(f"PR created: {pr_data['html_url']}")
            break
    else:
        # Clean up the remote branch
        subprocess.check_call(["git", "push", "origin", "--delete", PR_BRANCH_NAME])
        print("Failed to create PR", file=sys.stderr)
        sys.exit(1)

    resp = requests.post(
        f"https://api.github.com/repos/{OWNER}/{REPO}/pulls/{pr_data['number']}/requested_reviewers",
        headers={"Authorization": f"token {GITHUB_TOKEN}"},
        json={"reviewers": ["harupy"]},
    )
    resp.raise_for_status()


if __name__ == "__main__":
    main()

File: .github/workflows/sync.js
const { execSync } = require("child_process");

const shouldSync = (comment) => {
  return comment.body.trim() === "/sync";
};

const getPullInfo = async (context, github) => {
  const { owner, repo } = context.repo;
  const pull_number = context.issue.number;
  const pr = await github.rest.pulls.get({ owner, repo, pull_number });
  const {
    sha: head_sha,
    ref: head_ref,
    repo: { full_name },
  } = pr.data.head;
  const { sha: base_sha, ref: base_ref, repo: base_repo } = pr.data.base;
  return {
    repository: full_name,
    pull_number,
    head_sha,
    head_ref,
    base_sha,
    base_ref,
    base_repo: base_repo.full_name,
    author_association: pr.data.author_association,
    maintainer_can_modify: pr.data.maintainer_can_modify,
    mergeable: pr.data.mergeable,
    mergeable_state: pr.data.mergeable_state,
  };
};

const createInitialReaction = async (context, github) => {
  const { owner, repo } = context.repo;
  const { id: comment_id } = context.payload.comment;
  const { runId } = context;

  // Add rocket reaction
  await github.rest.reactions.createForIssueComment({
    owner,
    repo,
    comment_id,
    content: "rocket",
  });

  // Add workflow run link with pending emoji
  const workflowRunUrl = `https://github.com/${owner}/${repo}/actions/runs/${runId}`;
  const body = `⏳ [Sync workflow started](${workflowRunUrl})`;

  const response = await github.rest.issues.createComment({
    owner,
    repo,
    issue_number: context.issue.number,
    body,
  });

  return response.data.id;
};

const isAuthorAllowed = ({ author_association, user }) => {
  return (
    ["owner", "member", "collaborator"].includes(author_association.toLowerCase()) ||
    // Allow Copilot and mlflow-app bot to run this workflow
    (user &&
      user.type.toLowerCase() === "bot" &&
      ["copilot", "mlflow-app[bot]"].includes(user.login.toLowerCase()))
  );
};

const validateAuthorPermissions = async (context, github, core) => {
  const { comment } = context.payload;
  const { owner, repo } = context.repo;
  const { runId } = context;

  if (
    !isAuthorAllowed({
      author_association: comment.author_association,
      user: comment.user,
    })
  ) {
    const workflowRunUrl = `https://github.com/${owner}/${repo}/actions/runs/${runId}`;
    const message = `❌ **Sync failed**: Only repository owners, members, or collaborators can use the /sync command. @${comment.user.login} (${comment.author_association}) does not have sufficient permissions.

**Details:** [View workflow run](${workflowRunUrl})`;

    await github.rest.issues.createComment({
      owner,
      repo,
      issue_number: context.issue.number,
      body: message,
    });

    core.setFailed(`User ${comment.user.login} does not have sufficient permissions`);
  }
};

const validatePRConditions = async (context, github, pullInfo) => {
  const { owner, repo } = context.repo;
  const { runId } = context;
  const workflowRunUrl = `https://github.com/${owner}/${repo}/actions/runs/${runId}`;

  // Check if it's a fork PR
  const isForkPR = pullInfo.repository !== pullInfo.base_repo;

  // For fork PRs, check if maintainers can modify
  if (isForkPR && !pullInfo.maintainer_can_modify) {
    const message = `❌ **Sync failed**: For fork PRs, the "Allow edits and access to secrets by maintainers" checkbox must be checked.

Please:
1. Check the "Allow edits and access to secrets by maintainers" checkbox on this pull request
2. Comment \`/sync\` again

**Details:** [View workflow run](${workflowRunUrl})`;

    await github.rest.issues.createComment({
      owner,
      repo,
      issue_number: context.issue.number,
      body: message,
    });

    throw new Error("Fork PR does not allow maintainer edits");
  }

  // Check for merge conflicts
  if (pullInfo.mergeable === false) {
    const message = `❌ **Sync failed**: This PR has merge conflicts that must be resolved before syncing.

Please resolve the conflicts and then comment \`/sync\` again.

**Details:** [View workflow run](${workflowRunUrl})`;

    await github.rest.issues.createComment({
      owner,
      repo,
      issue_number: context.issue.number,
      body: message,
    });

    throw new Error("PR has merge conflicts");
  }

  // Check if mergeable state is unknown (GitHub is still computing)
  if (pullInfo.mergeable === null) {
    // Wait a bit and check again
    await new Promise((resolve) => setTimeout(resolve, 5000));

    const updatedPR = await github.rest.pulls.get({
      owner,
      repo,
      pull_number: pullInfo.pull_number,
    });

    if (updatedPR.data.mergeable === false) {
      const message = `❌ **Sync failed**: This PR has merge conflicts that must be resolved before syncing.

Please resolve the conflicts and then comment \`/sync\` again.

**Details:** [View workflow run](${workflowRunUrl})`;

      await github.rest.issues.createComment({
        owner,
        repo,
        issue_number: context.issue.number,
        body: message,
      });

      throw new Error("PR has merge conflicts");
    }
  }
};

const performSync = async (context, github, pullInfo, botToken) => {
  const { owner, repo } = context.repo;

  try {
    // Configure git
    execSync('git config user.name "mlflow-app[bot]"');
    execSync('git config user.email "mlflow-app[bot]@users.noreply.github.com"');

    // We're already on the PR head branch, just need to fetch and merge base
    const baseRepoUrl = `https://github.com/${pullInfo.base_repo}.git`;

    console.log(`Setting up base remote and fetching base branch`);

    // Clean up any existing remote
    try {
      execSync("git remote remove base", { stdio: "ignore" });
    } catch (e) {
      /* ignore */
    }

    execSync(`git remote add base ${baseRepoUrl}`);
    execSync(`git fetch base ${pullInfo.base_ref}`);

    console.log(`Merging base branch: ${pullInfo.base_ref}`);
    execSync(`git merge base/${pullInfo.base_ref} --no-edit`);

    console.log(`Pushing updated branch`);
    execSync(`git push origin HEAD`);

    console.log("Sync completed successfully");
  } catch (error) {
    console.error("Sync failed:", error.message);

    const { runId } = context;
    const workflowRunUrl = `https://github.com/${owner}/${repo}/actions/runs/${runId}`;
    const message = `❌ **Sync failed**: An error occurred during the sync process.

Error: \`${error.message}\`

**Details:** [View workflow run](${workflowRunUrl})`;

    await github.rest.issues.createComment({
      owner,
      repo,
      issue_number: context.issue.number,
      body: message,
    });

    throw error;
  }
};

const updateCommentStatus = async (context, github, success, workflowCommentId) => {
  const { owner, repo } = context.repo;
  const { runId } = context;
  const workflowRunUrl = `https://github.com/${owner}/${repo}/actions/runs/${runId}`;

  if (workflowCommentId) {
    const emoji = success ? "✅" : "❌";
    const status = success ? "completed successfully" : "failed";
    const updatedBody = `${emoji} [Sync workflow ${status}](${workflowRunUrl})`;

    try {
      await github.rest.issues.updateComment({
        owner,
        repo,
        comment_id: workflowCommentId,
        body: updatedBody,
      });
    } catch (error) {
      console.error("Failed to update workflow comment:", error.message);
    }
  }

  // If successful, add a success message
  if (success) {
    const message = `✅ **Sync completed**: The PR branch has been successfully updated with the latest changes from the base branch.

**Details:** [View workflow run](${workflowRunUrl})`;

    await github.rest.issues.createComment({
      owner,
      repo,
      issue_number: context.issue.number,
      body: message,
    });
  }
};

module.exports = {
  shouldSync,
  getPullInfo,
  createInitialReaction,
  validateAuthorPermissions,
  validatePRConditions,
  performSync,
  updateCommentStatus,
};
